pandas/core/dtypes/cast.py

```diff
@@ -264,9 +289,17 @@ def maybe_promote(dtype, fill_value=np.nan):
 
     # returns tuple of (dtype, fill_value)
     if issubclass(dtype.type, np.datetime64):
-        fill_value = tslibs.Timestamp(fill_value).value
```

Comment:
@jreback 
> > fixes a regression that was silently introduced in #23796 due to lack of tests

> what was the regression?

The try/catch around these were removed due to https://github.com/pandas-dev/pandas/pull/23796/commits/ecfe8248a856594aaa3d10c758e1b724ba2e2024#r235838683, but the Timestamp/Timedelta constructor choke when fed with strings/objects.

Reply:
what did this break from a user POV? if it didn't then don't revert this, it *should* fail.

Reply:
I didn't say it broke anything from a user perspective, just that #23796 broke something that wasn't tested (namely, trying to upcast from `datetime/timedelta` to string/object

Reply:
so these should probably be using: ``is_datetime64_dtype(dtype)`` and ``is_timedelta64_dtype(dtype)`` as more consistent with the rest of the codebase.

I am also not sure whether much of this routine (maybe_promote) is actually used in a meaningful way. Hence I don't think these checks are needed currently.

Reply:
> I am also not sure whether much of this routine (maybe_promote) is actually used in a meaningful way. Hence I don't think these checks are needed currently.

There's a list of places it appear in in #23833, but one of the key points is in `maybe_upcast_putmask`, which I want to use as the backend for `.update` in #23192, see #23606, #22358 etc.